Modules/_abc.c

```diff
@@ -0,0 +1,894 @@
+                           0, PyList_GET_SIZE(gs->pending), NULL);
+}
+
+PyDoc_STRVAR(_reset_registry_doc,
+"Internal ABC helper to reset registry of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_registry(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    if (!PyArg_UnpackTuple(args, "_reset_registry", 1, 1, &self)) {
+        return NULL;
+    }
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_registry->data) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(_reset_caches_doc,
+"Internal ABC helper to reset both caches of a given class.\n\
+\n\
+Should be only used by refleak.py");
+
+static PyObject *
+_reset_caches(PyObject *m, PyObject *args)
+{
+    PyObject *self;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_reset_caches", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (PySet_Clear(impl->_abc_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    /* also the second cache */
+    if (PySet_Clear(impl->_abc_negative_cache) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+    Py_RETURN_NONE;
+}
+
+PyDoc_STRVAR(_get_dump_doc,
+"Internal ABC helper for cache and registry debugging.\n\
+\n\
+Return shallow copies of registry, of both caches, and\n\
+negative cache version. Don't call this function directly,\n\
+instead use ABC._dump_registry() for a nice repr.");
+
+static PyObject *
+_get_dump(PyObject *m, PyObject *args)
+{
+    PyObject *self, *registry, *cache, *negative_cache;
+    _abc_data *impl;
+    if (!PyArg_UnpackTuple(args, "_get_dump", 1, 1, &self)) {
+        return NULL;
+    }
+    impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    registry = PyObject_CallMethod(impl->_abc_registry->data, "copy", NULL);
+    if (registry == NULL) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    cache = PyObject_CallMethod(impl->_abc_cache, "copy", NULL);
+    if (cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        return NULL;
+    }
+    negative_cache = PyObject_CallMethod(impl->_abc_negative_cache, "copy", NULL);
+    if (negative_cache == NULL) {
+        Py_DECREF(impl);
+        Py_DECREF(registry);
+        Py_DECREF(cache);
+        return NULL;
+    }
+    Py_INCREF(impl->_abc_negative_cache_version); /* PyTuple_Packdoesn't do this. */
+    PyObject *res = PyTuple_Pack(4,
+            registry, cache, negative_cache, impl->_abc_negative_cache_version);
+    Py_DECREF(impl);
+    return res;
+}
+
+// Compute set of abstract method names.
+static int
+compute_abstract_methods(PyObject *self)
+{
+    int ret = -1;
+    PyObject *abstracts = PyFrozenSet_New(NULL);
+    if (!abstracts) {
```

Comment:
nitpicking: I prefer explicit NULL checks: `if (abstracts == NULL)`.  I think we mostly do it this way in CPython and it's more readable this way.